src/adcs_handler.c

```diff
@@ -46,242 +50,450 @@ ADCS_returnState adcs_telecommand(uint32_t * command, uint32_t length) {
+    command[2] = overwrite_flag;
+    return adcs_telecommand(command, 3);
+  }
 }
 
-/**
- * @brief
- * 		Finalize Uploaded File Block
- * @param file_dest
- * 		File Destination
- * @param offset
- * 		Offset into file
- * @param block_length
- * 		Length of block
- * @return
- * 		<if applicable, brief description of what is returned by the function>
- */
-ADCS_returnState ADCS_finalize_upload_block(File_Upload_Dest file_dest, uint32_t offset, uint16_t block_length){
-  uint32_t command[4];
-  command[0] = FINALIZE_UPLOAD_BLOCK_ID;
-  command[1] = file_dest;
-  command[2] = offset;
-  command[3] = block_length;
-  adcs_telecommand(command, 4);
-}
+/*************************** BootLoader TMs ***************************/
 
 /**
  * @brief
- * 		Reset HoleMap for Upload Block
+ * 		Get BootLoader state
+ * @param uptime
+ * @param 11 boolean flags from Table 69
+ * 		sram1, sram2, sram_latch_not_recovered, sram_latch_recovered,
+ * sd_initial_err, sd_read_err, sd_write_err, external_flash_err,
+ * internal_flash_err, eeprom_err, bad_boot_reg, comms_radio_err
+ * @attention
+ * 		May be replaced with an array of flags
  * @return
- * 		<if applicable, brief description of what is returned by the function>
+ * 		Success of function defined in adcs_types.h
  */
-ADCS_returnState ADCS_reset_upload_block(){
-  uint32_t command = RESET_UPLOAD_BLOCK_ID;
-  return adcs_telecommand(&command, 1);
+ADCS_returnState ADCS_get_bootloader_state(uint16_t* uptime,
+                                           uint8_t* flags_arr) {
+  uint8_t telemetry[6];
+  ADCS_returnState state;
+  state = adcs_telemetry(GET_BOOTLOADER_STATE_ID, telemetry, 6);
+  *uptime = (telemetry[1] << 8) + telemetry[0];
+  uint32_t flags;
+  memcpy(&flags, &telemetry[2], 4);
+  for (int i = 0; i < 11; i++) {
+    *(flags_arr + i) = (flags >> i) & 1;
+  }
+  return state;
 }
 
 /**
  * @brief
- * 		Reset File List Read Pointer
+ * 		Get program information
+ * @param index
+ * 		Table 66 - Source Program index
+ * @param busy
+ * 		program busy reading
+ * @param file_size
+ * 		File size in bytes
+ * @param crc16_checksum
  * @return
- * 		<if applicable, brief description of what is returned by the function>
+ * 		Success of function defined in adcs_types.h
  */
-ADCS_returnState ADCS_reset_file_list_read_pointer(){
-  uint32_t command = RESET_FILE_LIST_READ_POINTER_ID;
-  return adcs_telecommand(&command, 1);
+ADCS_returnState ADCS_get_program_info(uint8_t* index, bool* busy,
+                                       uint32_t* file_size,
+                                       uint16_t* crc16_checksum) {
+  uint8_t telemetry[8];
+  ADCS_returnState state;
+  state = adcs_telemetry(GET_PROGRAM_INFO_ID, telemetry, 8);
+  *index = telemetry[0];
+  *busy = telemetry[1] & 1;  //*
+  // Different endianness! Should we use conversion functions?
+  *file_size = (telemetry[5] << 24) + (telemetry[4] << 16) +
+               (telemetry[3] << 8) + telemetry[2];
```

Comment:
Yeah, might be efficient to have a function to convert from little-endian, if that's what you're asking? Is there a function in the C library for that already?

Reply:
Yeah, I guess I should come up with something that checks that. We have the same thing in the service files.